app/services/dqt_record_check.rb

```diff
@@ -0,0 +1,126 @@
+# frozen_string_literal: true
+
+class DqtRecordCheck < ::BaseService
+  CheckResult = Struct.new(
+    :dqt_record,
+    :trn_matches,
+    :name_matches,
+    :dob_matches,
+    :nino_matches,
+    :total_matched,
+    :failure_reason,
+  )
+
+  def call
+    if magic_date_criteria_met?
+      return magic_response
+    end
+
+    check_record
+  end
+
+private
+
+  attr_reader :trn, :nino, :full_name, :date_of_birth, :check_first_name_only
+  alias_method :check_first_name_only?, :check_first_name_only
+
+  def initialize(trn:, full_name:, date_of_birth:, nino: nil, check_first_name_only: true)
+    @trn = trn
+    @full_name = full_name
+    @date_of_birth = date_of_birth
+    @nino = nino
+    @check_first_name_only = check_first_name_only
+  end
+
+  def dqt_record(trn, nino)
+    full_dqt_client.get_record(trn:, birthdate: date_of_birth, nino:)
+  end
+
+  def full_dqt_client
+    @full_dqt_client ||= FullDQT::Client.new
+  end
+
+  def check_record
+    return check_failure(:trn_and_nino_blank) if trn.blank? && nino.blank?
+
+    @trn = "0000001" if trn.blank?
+
+    padded_trn = TeacherReferenceNumber.new(trn).formatted_trn
+    dqt_record = dqt_record(padded_trn, nino)
+
+    return check_failure(:no_match_found) if dqt_record.nil?
+    return check_failure(:found_but_not_active) if dqt_record["state_name"] != "Active"
+
+    matches = 0
+    trn_matches = padded_trn == dqt_record["trn"]
+    matches += 1 if trn_matches
+
+    name_matches = if check_first_name_only?
+                     full_name.split(" ").first.downcase == dqt_record["name"].split(" ").first.downcase
+                   else
+                     full_name.downcase == dqt_record["name"].downcase
+                   end
+
+    matches += 1 if name_matches
+
+    dob_matches = date_of_birth == dqt_record["dob"]
+    matches += 1 if dob_matches
+    nino_matches = nino.present? && nino.downcase == dqt_record["ni_number"]&.downcase
+    matches += 1 if nino_matches
+
+    if matches < 3 && (trn_matches && trn != "1")
+      # If a participant mistypes their TRN and enters someone else's, we should search by NINO instead
+      # The API first matches by (mandatory) TRN, then by NINO if it finds no results. This works around that.
+      @trn = "0000001"
+      return check_record
+    end
+
+    CheckResult.new(dqt_record, trn_matches, name_matches, dob_matches, nino_matches, matches)
+  end
```

Comment:
That was really the point of breaking this out of the `ParticipantValidationService` - grab the record and do the matching checks - but also allow us to return the DQT record and the check results to the caller.  Do you think this needs breaking down further? I'm not sure what value it would have outside of this?

Reply:
I've simplified this a bit now by adding a present for the returned dqt record